### PR TITLE
Per-format TextWatcher, pause text watcher for undo/redo

### PR DESCRIPTION
**Pull request by harshad1**
This is a follow up / bug fix for https://github.com/gsantner/markor/pull/883.

Instead of being attached to the HighlightingEditor in the constructor, the TextWatcher (ListHandler) is handled through the Highlighter (like the InputFilter) and only applied to the correct file format. This prevents the TextWatcher from making unwanted modifications for non-markdown file types.

In addition, the TextWatcher and Input filter are paused when undo and redo are performed. This prevents the automatic addition / deletion of things when undo and redo are used.

### CODE REVIEW

File: app/src/main/java/net/gsantner/markor/format/TextFormat.java [hunk 1]
```diff
@@ -26,6 +27,7 @@
 import net.gsantner.markor.model.Document;
 import net.gsantner.markor.ui.hleditor.Highlighter;
 import net.gsantner.markor.ui.hleditor.HighlightingEditor;
+import net.gsantner.markor.ui.hleditor.ListHandler;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/TextFormat.java:30]: Codacy Issue found: [Avoid unused imports such as 'net.gsantner.markor.ui.hleditor.ListHandler'](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 7
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883&bid=17649600)

File: app/src/main/java/net/gsantner/markor/format/TextFormat.java [hunk 2]
```diff
@@ -104,6 +106,7 @@ public static TextFormat getFormat(int formatId, Activity activity, Document doc
     private TextActions _textActions;
     private Highlighter _highlighter;
     private TextConverter _converter;
+    private TextWatcher _watcher = null;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/TextFormat.java:109]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883)

File: app/src/main/java/net/gsantner/markor/format/TextFormat.java (hunk 1 above)
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/TextFormat.java:30]: Codacy Issue found: [Unused import - net.gsantner.markor.ui.hleditor.ListHandler.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java [hunk 3]
```diff
@@ -11,6 +11,7 @@
 
 import android.content.Context;
 import android.os.Handler;
+import android.support.annotation.NonNull;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java:14]: Codacy Issue found: [Avoid unused imports such as 'android.support.annotation.NonNull'](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883)

File: app/src/main/java/net/gsantner/markor/format/TextFormat.java (hunk 2 above)
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/TextFormat.java:109]: Codacy Issue found: [Avoid unused private fields such as '_watcher'.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java (hunk 3 above)
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java:14]: Codacy Issue found: [Unused import - android.support.annotation.NonNull.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/Highlighter.java [hunk 4]
```diff
@@ -73,6 +74,7 @@ protected static Highlighter getDefaultHighlighter(HighlightingEditor hlEditor,
     protected boolean _highlightLinks = true;
     protected final boolean _highlightHexcolor;
     protected final Document _document;
+    private TextWatcher _modifier = null;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/Highlighter.java:77]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883&bid=17649600)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/Highlighter.java (hunk 4 above)
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/Highlighter.java:77]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883)

File: app/src/main/java/net/gsantner/markor/format/TextFormat.java [hunk 5]
```diff
@@ -10,6 +10,7 @@
 package net.gsantner.markor.format;
 
 import android.app.Activity;
+import android.text.TextWatcher;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/TextFormat.java:13]: Codacy Issue found: [Unused import - android.text.TextWatcher.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883)

Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/TextFormat.java:13]: Codacy Issue found: [Avoid unused imports such as 'android.text.TextWatcher'](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883&bid=17649600)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/Highlighter.java (hunk 4 above)
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/Highlighter.java:77]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883)

**Comment by gsantner** [comment]
Can the other PR be closed?

By the way, the DocumentIO history approach was my own, till I found `TextViewUndoRedo.java`. Basically that is the current history implementation and the old one is just living mostly unused in the codebase.

**Comment by harshad1** [comment]
> Can the other PR be closed?

Have closed other PR. It was mostly me experimenting with different parts of the undo/redo functionality. I will eventually open a new PR once I figure out a clean way to solve #894 .

> By the way, the DocumentIO history approach was my own, till I found `TextViewUndoRedo.java`. Basically that is the current history implementation and the old one is just living mostly unused in the codebase.

Got it.

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883&bid=17649600)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/Highlighter.java (hunk 4 above)
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/Highlighter.java:77]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java [hunk 6]
```diff
@@ -45,6 +47,7 @@ public boolean isCurrentLineEmpty() {
     private boolean _hlEnabled = false;
     private boolean _isSpellingRedUnderline;
     private Highlighter _hl;
+    private Set<TextWatcher> _appliedModifiers; /* Tracks currently applied modifiers */
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java:50]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883)

**Comment by gsantner** [comment]
@harshad1 
Is the PR ready? I found its not too hard to look over as is

File: app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java [hunk 7]
```diff
@@ -104,11 +107,11 @@ public void beforeTextChanged(CharSequence s, int start, int count, int after) {
             }
         });
 
-        // This watcher detects and handles termination of lists when last item is empty
-        addTextChangedListener(new ListHandler());
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java:108]: Make the variable final and init it directly with new hashset

**Comment by harshad1** [comment]
It's ready to go IMO.

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883&bid=17649600)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/Highlighter.java (hunk 4 above)
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/Highlighter.java:77]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java [hunk 8]
```diff
@@ -45,6 +47,7 @@ public boolean isCurrentLineEmpty() {
     private boolean _hlEnabled = false;
     private boolean _isSpellingRedUnderline;
     private Highlighter _hl;
+    private final Set<TextWatcher> _appliedModifiers = new HashSet<>(); /* Tracks currently applied modifiers */
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java:50]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5382883)

**Pull request merged**